### PR TITLE
[IMP] payment(_adyen): allow partial capture

### PR DESCRIPTION
**Pull request by Demesmaeker**
Before this commit, it was not possible to partially capture a
transaction from Odoo, and doing so in the provider backend would often
result in a full capture in Odoo when capture was supported.

With this commit, partial captures are made available in Odoo directly
from the sales order or invoice, for providers that support them.
Acquirer can either only support full capture or also support partial
ones. It also optionally managed the automatic void of the remaining
amount at the user request when multiple captures are supported by the
provider.

As of now, the only provider allowing partial capture is Adyen.

task-2728768

See also:
- https://github.com/odoo/enterprise/pull/35205
- https://github.com/odoo/documentation/pull/2063

### CODE REVIEW

File: addons/payment/wizards/payment_capture_wizard.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        string="Maximum Capture Allowed", compute='_compute_amount_available',
+    )
+    captured_amount = fields.Monetary(string="Already Captured", compute='_compute_captured_amount')
+    voided_amount = fields.Monetary(string="Already Voided", compute='_compute_voided_amount')
+    amount_to_capture = fields.Monetary(
+        string="Capture Amount", compute='_compute_amount_to_capture', store=True, readonly=False,
+    )
+    currency_id = fields.Many2one(string="Currency", related='transaction_ids.currency_id')
+    # TODO edm ask anv: is this okay with ids instead of id? We know it's the same, but what would
+    #  happens if it wasn't?
```
Comment (chevalierv) [inline addons/payment/wizards/payment_capture_wizard.py:31]:
```suggestion
```

**Review by chevalierv — CHANGES_REQUESTED** [review]
First Pass :duck: 

A bit of nitpicking ;) I didn't read all the files, just wanted to publish want I already read ^^

Btw it seems that the child transaction of transactions on invoices is linked to the documents but it's not the case for the sale orders. Can you check that?

File: addons/payment/models/payment_transaction.py [hunk 2]
```diff
@@ -278,8 +299,11 @@ def action_void(self):
 
         payment_utils.check_rights_on_recordset(self)
         for tx in self:
+            captured_amount = sum(child.amount for child in tx.child_transaction_ids.filtered(
+                lambda t: t.state == 'done' and t.operation == tx.operation
+            ))
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/models/payment_transaction.py:304]: Yup, could have go with a `t.operation != 'refund'`, but it's less maintainable with at some point we add another child operation =]
Done

File: addons/payment/models/payment_transaction.py [hunk 3]
```diff
@@ -756,6 +812,29 @@ def classify_by_state(transactions_):
...
+                    self.env, child_tx.source_transaction_id.amount, child_tx.currency_id
+                )
+                state_message = _(
+                    "The transaction with reference %(ref)s for %(amount)s has been confirmed "
+                    "following the processing of its capture and void children (%(pvd_name)s).",
+                    ref=child_tx.source_transaction_id.reference,
+                    amount=formatted_amount,
+                    pvd_name=child_tx.provider_id.name,
+                )
+                child_tx.source_transaction_id._update_state(('authorized',), 'done', state_message)
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/models/payment_transaction.py:836]: Hm, I understand why you don't like the fact I directly use the `_update_sate` method. For me, this `_update_source_transaction_state` is at same level of use that all the `_set_state` methods, so it doesn't bother me to use it like that.
Other solution could be to pass a new arg to `_set_done` like `_set_done(state_message, is_parent=True)` But that seems even messier to me. What do you think? :thinking:

File: addons/payment/wizards/payment_capture_wizard.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        string="Has Remaining Amount", compute='_compute_has_remaining_amount',
+    )
+    void_remaining_amount = fields.Boolean(string="Void remaining amount")
+
+    #=== COMPUTE METHODS ===#
+
+    @api.depends('transaction_ids')
+    def _compute_authorized_amount(self):
+        for wizard in self:
+            wizard.authorized_amount = sum(wizard.transaction_ids.mapped('amount'))
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard.py]: Nop, multiple partial captures, the source tx might have child but still needs to be counted in =]
Or even being fully processed in children, being in state "done", and still needs to be counted in the authorized amount

**Comment by Demesmaeker** [comment]
@chevalierv 
I took off the code that added the invoice_ids to the child transactions instead of adding the sale orders to the other child tx as you suggested yesterday. I know, at first, doing the opposite seems better, but this is why:

- Tx are highly technical object, child tx even more, usually, the users don't have to go there, so having the information isn't that necessary on the child tx
- In the case the user wants the info, he can easily go to the source transaction which is linked to the invoice, so...
- From the so/invoice, having only the parent transactions linked is easier code wise, when we want to get the information about what was done etc, and ensure we don't get the noise from the child tx, but could still get them since they are linked to their parent tx.
- Adding it means making some overrides that needs to be maintained too
- Adding it also means ensure everywhere we took the tx from a so/invoice that we only took those that are not child.

tldr; having info of so/invoice on child tx is overkill and brings a lot of potential coding issues we need to be aware of everytime

File: addons/payment/wizards/payment_capture_wizard.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard.py:48]: And as it is, only parent transaction are passed in the wizard, so that's not an issue since I took off the `invoices_ids` on the child tx =]

[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/payment/wizards/payment_capture_wizard.py:48]: Indeed, my suggestion should have been:
```suggestion
            wizard.authorized_amount = sum(wizard.transaction_ids.filter(lamba tx: not tx.source_transaction_ids).mapped('amount'))
```

But it's okay like this since it should not be possible to get child transactions in this wizard, and as we discussed, it will be a good way to find out if there is a problem in that flow

**Review by chevalierv — CHANGES_REQUESTED** [review]
Last of my nitpicking comments :)

There is still awkward behavior with the invoices. Now, I don't see the confirmation that Adyen captured the transactions in the chatter and I expect to see the child transactions under the total amount in the form view.

File: addons/payment/wizards/payment_capture_wizard.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    def _compute_captured_amount(self):
+        for wizard in self:
+            captured_src_tx_no_child = wizard.transaction_ids.filtered(
+                lambda tx: tx.state == 'done' and not tx.child_transaction_ids
+            )
+            captured_child_tx = wizard.transaction_ids.child_transaction_ids.filtered(
+                lambda tx: tx.state == 'done'
+            )
+            captured_tx = captured_src_tx_no_child | captured_child_tx
+            wizard.captured_amount = sum(tx.amount for tx in captured_tx)
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard.py:60]: Ha, nice, yes, indeed =D

[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard.py:60]: Didn't choose your suggestion, but changed it anyway to something else as it was indeed unclear =]

File: addons/payment/wizards/payment_capture_wizard.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    def _compute_amount_to_capture(self):
+        """ Set the default amount to capture to the amount available for capture. """
+        for wizard in self:
+            wizard.amount_to_capture = wizard.amount_available
+
+    def _compute_support_partial_capture(self):
+        for wizard in self:
+            wizard.support_partial_capture = not any(
+                tx.provider_id.support_manual_capture != 'partial' for tx in wizard.transaction_ids
+            )
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard.py:88]: :facepalm: yup, indeed

File: addons/sale/models/sale_order.py [hunk 7]
```diff
@@ -1287,14 +1287,15 @@ def _force_lines_to_invoice_policy_order(self):
 
     def payment_action_capture(self):
         """ Capture all transactions linked to this sale order. """
+        self.ensure_one()
         payment_utils.check_rights_on_recordset(self)
-        # In sudo mode because we need to be able to read on provider fields.
-        self.authorized_transaction_ids.sudo().action_capture()
+        # In sudo mode because we need to bypass the checks on the rights in the transactions.
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/sale/models/sale_order.py:1292]: Useless diff, I could take that off as this isn't related to this PR. Rewrote that when I investigated the fact that it's to bypass the check made on `action_capture`/`action_void`, more than to read on the provider field (which indeed need the sudo, but that's not the question here, since the sudo is written and commented in the `action_capture`/`action_void` flow)

**Review by chevalierv — APPROVED** [review]

File: addons/payment_adyen/models/payment_transaction.py [hunk 8]
```diff
@@ -375,17 +435,27 @@ def _process_notification_data(self, notification_data):
...
-                    _("The capture of the transaction with reference %s failed.", self.reference)
-                )
+                if self.source_transaction_id:  # child_tx => the event can't be retried
+                    self._set_error(_(
+                        "The capture of the transaction with reference %s failed.", self.reference
+                    ))
+                else:  # source tx with failed capture stays in its state, could be captured again
+                    self._log_message_on_linked_documents(_(
+                        "The capture of the transaction with reference %s failed.", self.reference
+                    ))
```
Comment (AntoineVDV) [inline addons/payment_adyen/models/payment_transaction.py:458]: It could be worth merging all these messages into two variables to save some space.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Great task! :)

Only a "few" minor points to address before merging :point_down:

File: addons/payment/models/payment_provider.py [hunk 9]
```diff
@@ -163,8 +163,10 @@ def _valid_field_parameter(self, field, name):
     support_tokenization = fields.Boolean(
         string="Tokenization Supported", compute='_compute_feature_support_fields'
     )
-    support_manual_capture = fields.Boolean(
-        string="Manual Capture Supported", compute='_compute_feature_support_fields'
+    support_manual_capture = fields.Selection(
+        string="Manual Capture Supported",
+        selection=[('full_only', "Full Only"), ('partial', "Partial")],
+        compute='_compute_feature_support_fields',
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/models/payment_provider.py:169]: For a compute not stored? :thinking:

File: addons/payment/models/payment_transaction.py [hunk 10]
```diff
@@ -70,20 +70,20 @@ def _lang_get(self):
...
         string="Source Transaction",
         comodel_name='payment.transaction',
-        help="The source transaction of related refund transactions",
+        help="The source transaction of related child transactions",
         readonly=True,
     )
     child_transaction_ids = fields.One2many(
         string="Child Transactions",
-        help="The child transactions of the source transaction.",
+        help="The child transactions of the related source transaction.",
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/models/payment_transaction.py:86]: Yes, indeed, guess I changed that string one too many times '^^

File: addons/payment/wizards/payment_capture_wizard.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        for wizard in self:
+            wizard.support_partial_capture = all(
+                tx.provider_id.support_manual_capture == 'partial' for tx in wizard.transaction_ids
+            )
+
+    @api.depends('transaction_ids')
+    def _compute_has_draft_children(self):
+        for wizard in self:
+            draft_children = wizard.transaction_ids.child_transaction_ids.filtered(
+                lambda tx: tx.state == 'draft'
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard.py:99]: Children will never be in a pending state, they will always go from draft to either done or cancel. Looks like excessive caution to add it here too :thinking:

File: addons/payment/wizards/payment_capture_wizard_views.xml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                    <span>
+                        <label for="amount_to_capture" class="oe_inline"/>
+                    </span>
+                    <span>
+                        <field name="amount_to_capture" class="oe_inline"
+                               attrs="{'readonly': [('support_partial_capture', '=', 'full_only')]}"/>
+                        <i class="fa fa-info-circle oe_inline"
+                           title="To change the amount you want to capture, all the transactions must come from providers which supports the partial transaction. If that's not the case but you intended to do a partial capture, manage it directly from the transaction which allows it."/>
+                    </span>
+                    <field name="void_remaining_amount" attrs="{'readonly': [('has_remaining_amount', '=', False)]}"/>
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard_views.xml:44]: It was discussed with Dala, and we decided to always show it because the user needs to click somewhere else for this field to appears/disappear whenever it's necessary.

File: addons/payment/wizards/payment_capture_wizard_views.xml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+                    <field name="transaction_ids" invisible="1"/>
+                    <field name="currency_id" invisible="1"/>
+                    <field name="is_amount_allowed" invisible="1"/>
+                    <field name="authorized_amount"/>
+                    <field name="captured_amount"
+                           attrs="{'invisible': [('captured_amount', '&lt;=', 0)]}"/>
+                    <field name="voided_amount"
+                           attrs="{'invisible': [('voided_amount', '&lt;=', 0)]}"/>
+                </group>
+                <hr/><br/>
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_capture_wizard_views.xml:32]: It's better when the wizard is basic, indeed, but one we add `already captured amount` and or `already voided amount`, alert boxes... I find it quite messy, that's why I added the separation line. But I can delete it if you'd prefer =]

File: addons/account_payment/models/payment_transaction.py [hunk 14]
```diff
@@ -118,7 +121,11 @@ def _reconcile_after_done(self):
         self.invoice_ids.filtered(lambda inv: inv.state == 'draft').action_post()
 
         # Create and post missing payments for transactions requiring reconciliation
-        for tx in self.filtered(lambda t: t.operation != 'validation' and not t.payment_id):
+        for tx in self.filtered(
+            lambda t: t.operation != 'validation'
+                and not t.payment_id
+                and not t.child_transaction_ids.filtered(lambda t: t.state in ['done', 'cancel'])
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/account_payment/models/payment_transaction.py:127]: I choose another cleaner version (IMO)

File: addons/account_payment/models/payment_transaction.py [hunk 15]
```diff
@@ -186,8 +198,10 @@ def _log_message_on_linked_documents(self, message):
         """
         self.ensure_one()
         self = self.with_user(SUPERUSER_ID)  # Log messages as 'OdooBot'
-        if self.source_transaction_id.payment_id:
-            self.source_transaction_id.payment_id.message_post(body=message)
+        if self.source_transaction_id:
+            payment_id = self.source_transaction_id.payment_id
+            if payment_id:
+                payment_id.message_post(body=message)
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/account_payment/models/payment_transaction.py:204]: Hm, no, because we still need to post the message to the `source_tx`'s invoices if there isn't a `payment_id`.

I put the `for` loop before the `if` to make it clearer =]

We didn't put that with the refund feature because before, we linked the invoices to the refund tx, which isn't the case anymore.

File: addons/payment_adyen/wizards/payment_capture_wizard_views.xml [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    <record id="payment_capture_wizard_view_form" model="ir.ui.view">
+        <field name="name">payment.adyen.capture.wizard.form</field>
+        <field name="model">payment.capture.wizard</field>
+        <field name="inherit_id" ref="payment.payment_capture_wizard_view_form"/>
+        <field name="arch" type="xml">
+            <div id="alert_draft_capture_tx" position='before'>
+                <field name="has_adyen_tx" invisible="1"/>
+                <div id="alert_delayed_capture_tx"
+                     role="alert"
+                     class="alert alert-warning"
```
Comment (AntoineVDV) [inline addons/payment_adyen/wizards/payment_capture_wizard_views.xml:13]: This is not a warning, but rather a note.
```suggestion
                     class="alert alert-info"
```

File: addons/payment_adyen/wizards/payment_capture_wizard_views.xml [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+                     attrs="{'invisible': [('has_adyen_tx', '=', False)]}">
+                    <p>
+                        <strong>Warning!</strong> The capture or void of the transaction might take
```
Comment (AntoineVDV) [inline addons/payment_adyen/wizards/payment_capture_wizard_views.xml:16]:
```suggestion
                        The capture or void of the transaction might take
```

File: addons/payment_adyen/wizards/payment_capture_wizard_views.xml [hunk 18]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="payment_capture_wizard_view_form" model="ir.ui.view">
+        <field name="name">payment.adyen.capture.wizard.form</field>
+        <field name="model">payment.capture.wizard</field>
+        <field name="inherit_id" ref="payment.payment_capture_wizard_view_form"/>
+        <field name="arch" type="xml">
+            <div id="alert_draft_capture_tx" position='before'>
```
Comment (AntoineVDV) [inline addons/payment_adyen/wizards/payment_capture_wizard_views.xml:9]:
```suggestion
            <div id="alert_draft_capture_tx" position="before">
```

File: addons/payment/models/payment_transaction.py [hunk 19]
```diff
@@ -589,6 +605,42 @@ def _ensure_provider_is_not_disabled(self):
...
+        """ Create a new transaction with the current transaction as its parent transaction.
+
+        This happens only in case of a refund or a partial capture (where the initial transaction is
+        split between smaller transactions, either captured or voided).
+
+        Note: self.ensure_one()
+
+        :param float amount: The strictly positive amount of the child transaction, in the same
+                             currency as the source transaction.
+        :param bool is_refund: Whether the child transaction is a refund or not.
```
Comment (AntoineVDV) [inline addons/payment/models/payment_transaction.py:618]:
```suggestion
        :param bool is_refund: Whether the child transaction is a refund.
```

File: addons/payment/wizards/payment_capture_wizard_views.xml [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+                <hr/>
+                <group name="input_fields">
+                    <label for="amount_to_capture" class="oe_inline"/>
+                    <div class="o_row">
+                        <field name="amount_to_capture"
+                               class="oe_inline"
+                               attrs="{'readonly': [('support_partial_capture', '=', 'full_only')]}"/>
+                        <i class="fa fa-info-circle oe_inline"
+                           attrs="{'invisible': [('support_partial_capture', '!=', 'full_only')]}"
+                           title="Some of the transaction you intend to capture can only be captured in full. Handle the transactions individually to capture a partial amount."/>
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_capture_wizard_views.xml:41]:
```suggestion
                           title="Some of the transactions you intend to capture can only be captured in full. Handle the transactions individually to capture a partial amount."/>
```

File: addons/payment/wizards/payment_capture_wizard.py [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+                    self.env, wizard.available_amount, wizard.currency_id
+                )
+                raise ValidationError(_(
+                    "The amount to capture must be positive and cannot be superior to %s.",
+                    formatted_amount
+                ))
+            if not wizard.support_partial_capture \
+               and wizard.amount_to_capture != wizard.available_amount:
+                raise ValidationError(_(
+                    "Some of the transaction you intend to capture can only be captured in full. "
```
Comment (AntoineVDV) [inline addons/payment/wizards/payment_capture_wizard.py:123]:
```suggestion
                    "Some of the transactions you intend to capture can only be captured in full. "
```

**Review by AntoineVDV — APPROVED** [review]

**Comment by AntoineVDV** [comment]
@robodoo r+
